### ISSUE TITLE
Référent par fournisseur plutôt que par produit ?

### ISSUE THREAD

**Issue by JocelynDelalande** [Jojo144/compteur_du_gase#81]
Je n'ai pas connaissance de groupements utilisant vraiment les référents à l'échelle du produit, du coup ça donne plutôt une lourdeur dans la gestion (il faut saisir/mettre à jour personne en référence sur tous les produits d'un fournisseur).

Au passage, il y a des cas où avoir plusieurs référent·es à recevoir les alertes serait pertinent.

**Comment by Jojo144** [comment]
Exemple : Camille est référent.e du riz qu'on achète à une coopérative de riz, sauf des fois à Terra Libra pour faire la jonction entre les commandes. Du coup c'est bien s'iel est référente du riz Terra Libra même si iel n'est pas référente Terra Libra... Un peu tiré par les cheveux ? Bon d'accord.

Tout à fait d'accord avec le passage. Ne serait ce que 2 ou 3 max ça serait mieux que 1.

**Comment by JocelynDelalande** [comment]
Plan pour la dépréciation progressive de la notion de réf produit au profit de la réf fournisseur.

## Ajout de la notion de réf fournisseur

1. Champ many2many qui lie Fournisseur et Personne, car on veut éventuellement plusieurs réf fournisseur
2. Migration : on « élit » un réf fournisseur en utilisant les réf actif : l'user actif qui est réf pour le + de produits chez un fournisseur. on initialise ainsi les réf fournisseur en base.
3. Si le réf fournisseur était réf  de 100% des produits actifs du fournisseur, on supprime les références produits automatiquement, sinon on les laisse. 
4. Branchement du champ réf foruisseur aux alertes stock

## Dépréciation progressive

Proposition : suppression au bout d'un an de la fonction de réf produits si pas de retour (via mail compteur@girole), jusque là on garde les 2 notions qui cohabitent et on incite à supprimer les réfs produits (pour éviter les alertes en double).

1. Création d'une page sur le compteur qui explique le tout
2. On met un warning, avec lien vers la page explicative aux endroits suivants : 
  - dans les alertes stock par mail
  - sur la page produit des produits qui possèdent un réf produit

## Au bout d'1 an : suppression de la fonction réf produit (ou pas)

Si pas de retour : 

- suppression des réfs produit en BDD
- suppression de la page d'explication
- suppression du code lié à la dépréciation progressive

Si retours souhaitant conserver cette notion : on réfléchit comment articuler les deux de manière gérable.

**Comment by Jojo144** [comment]
Ok super ! Ça fait un peu de boulot !

**Comment by ghostinthemuf** [comment]
> Migration : on « élit » un réf fournisseur en utilisant les réf actif : l'user actif qui est réf pour le + de produits chez un fournisseur. on initialise ainsi les réf fournisseur en base.
> Si le réf fournisseur était réf de 100% des produits actifs du fournisseur, on supprime les références produits automatiquement, sinon on les laisse.

Puisque la fonctionnalité le permettra, pourquoi, dans le cas où les produits sont répartis entre plusieurs référents pour un même fournisseur, ne pas initialiser autant de référents fournisseurs ? A chaud j'ai l'impression qu'un doublonnage de l'info (= 2 ou plusieurs utilisateurs recevront une même alerte) est moins gênant qu'une perte d'info (= un utilisateur ne recevra plus des alertes qu'il recevait précédemment).
Et ça permettrait de supprimer directement toutes les refs produits.

**Comment by JocelynDelalande** [comment]
> A chaud j'ai l'impression qu'un doublonnage de l'info (= 2 ou plusieurs utilisateurs recevront une même alerte) est moins gênant qu'une perte d'info (= un utilisateur ne recevra plus des alertes qu'il recevait précédemment).
> Et ça permettrait de supprimer directement toutes les refs produits.

Il n'y a pas de perte d'infos immédiate, mais quand/si la feature est virée, c'est vrai, ça fera ça. Ok je fais comme tu dis :-)

**Comment by JocelynDelalande** [comment]
> > Migration : on « élit » un réf fournisseur en utilisant les réf actif : l'user actif qui est réf pour le + de produits chez un fournisseur. on initialise ainsi les réf fournisseur en base.
> > Si le réf fournisseur était réf de 100% des produits actifs du fournisseur, on supprime les références produits automatiquement, sinon on les laisse.
> 
> Puisque la fonctionnalité le permettra, pourquoi, dans le cas où les produits sont répartis entre plusieurs référents pour un même fournisseur, ne pas initialiser autant de référents fournisseurs ? A chaud j'ai l'impression qu'un doublonnage de l'info (= 2 ou plusieurs utilisateurs recevront une même alerte) est moins gênant qu'une perte d'info (= un utilisateur ne recevra plus des alertes qu'il recevait précédemment). Et ça permettrait de supprimer directement toutes les refs produits.

@ghostinthemuf  à la réflexion je pense que c'est moins bien : Appelons ta proposition **B** et la mienne **A**.

Avec **A**. on détecte à grosse maille qui a potentiellement un usage réel des « référents produit » : c'est à dire plusieurs personnes qui se répartissent les produits d'un fournisseur. Et on va pouvoir afficher des warnings uniquement dans ces cas là. Par ex pour le GASE à roulettes, on pourra ne pas afficher de warning de dépréciation (sauf si les réfs produits sont mal renseignés). Et les users peuvent agir sur ces warnings pour les faire disparaître : il faut supprimer les référents par produits.

Avec **B**. on ne détecte rien et on sera obligés de « polluer » de warnings de dépréciation tout le monde tout le temps.

**Comment by ghostinthemuf** [comment]
C'est vrai que je n'avais pas pensé à la gestion des warnings. En fait ce qui me gêne avec **A.**, c'est qu'on prend des décisions à la place des utilisateurs (dont certains vont par ex. se retrouver référents pour des produits qu'ils ne géraient pas avant ça). Mais bon, on se prend sûrement la tête pour pas grand chose ^^ **A.** me convient aussi. Mais ça me fait quand même me demander si on ne devrait pas aussi communiquer en amont, avant les warnings ? Il s'agit quand même d'une évolution fonctionnelle pas anodine.

**Comment by JocelynDelalande** [comment]
Oui, on peut communiquer amont dans la mesure des contacts que l'on a (tu as ça toi @Jojo144  une liste de contacts ?).

PS: Pour B. on ajoute encore plus de personnes en référent·es sans leur consentement. Puisqu'être réf  sur un seul produit suffit pour être promu référent fournisseur.

**Comment by JocelynDelalande** [comment]
Est-ce que le texte suivant est clair pour la page de warning ? Retours bienvenus.

Ça pourrait servir de base à un mail envoyé aux gases connus : 


> # Disparition prévue des référent·es par produit au profit des référent·es par fournisseur au 1/6/2027
> Le compteur du GASE possède une notion de _référent par produit_. Le référent reçoit par mail des notifs quand le stock d'un produit passe sous le seuil d'alerte.
> 
> Il semble que cette notion recouvre assez mal les usages des GASEs, d'une part parce-que le bon endroit pour gérer des référents est plutôt le fournisseur que le produit, et d'autre part car on souhaite pouvoir avoir possiblement *plusieurs* référent·es.
> 
> C'est pourquoi une notion de *référent·e par fournisseur* a été introduite récement. Pour l'instant, elle co-habite avec celle de référent·e produit. Les référent·es produit ET les référent·es fournisseur sont prévenues par mail d'un stock trop bas.
> 
> Puisque nous (équipe de développement du logiciel) n'avons pas a-priori d'exemple d'usage réel de la notion de référent prouit, nous prévoyons de **supprimer la notion de référent·es par produit au 1/6/2027**, sauf si…
> 
> Si la notion de référent·e produit conserve un usage dans ton GASE, contacte jojo144@girole.fr pour expliquer ton usage et défendre le sauvetage des référent·es produit. Nous pourrons alors envisager de faire cohabiter les deux notions sur le long terme :-).
>
> ## Comment paramétrer les nouveaux référents fournisseur ?
> Pour chaque fournisseur, l**a personne qui avait le plus de produits en référence a été « élue » référente fournisseur automatiquement** à la mise à jour. Cette personne a également été retirée de la référence des produits concernés pour éviter les notifications en double.  Les personnes qui étaient « référentes minoritaires » sont, elles, restées référentes de leur(s) produit(s).
>
> ## Comment se débarasser de ce warning ?
> Vu que la fonction est amenée à une probable disparition, l'existence de référent·es produits produit des warnings a différents endroits du logiciel. Supprimer les référents des produits (si ils sont effectivement indésirables) supprimera les warnings. Il est possible de passer les personnes co-référent·es fournisseur manuellement à la place (depuis la fiche fournisseur).

**Comment by JocelynDelalande** [comment]
Pour faire des retours sur le texte, vous pouvez le faire directement en commentaire sur le code de https://github.com/Jojo144/compteur_du_gase/pull/130
